commands/upCommand.go

```diff
@@ -0,0 +1,88 @@
+package commands
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net/http"
+	"strings"
+	"time"
+
+	"github.com/UCCNetsoc/discord-bot/embed"
+
+	"github.com/bwmarrin/discordgo"
+	"github.com/matryer/try"
+	"github.com/spf13/viper"
+)
+
+type statusCheck struct {
+	Site    string
+	Time    int64
+	Success bool
+	Error   error
+}
+
+// Up command to check the status of various websites hosted on Netsoc servers
+func checkUpCommand(ctx context.Context, s *discordgo.Session, m *discordgo.MessageCreate) {
+	// Check if committee channel, don't allow in public server
+	if isCommittee(m) {
+		return
+	}
+
+	sites := strings.Split(viper.GetString("netsoc.sites"), ",")
+
+	// Run on a separate goroutine to not block bot
+	go checkStatuses(s, m, sites)
+}
+
+func checkStatuses(s *discordgo.Session, m *discordgo.MessageCreate, sites []string) {
+	// Create a channel to receive the status checks, whenever they complete
+	statuses := make(chan statusCheck)
+	// Run each status check on a separate goroutine as to not block each other
+	for _, site := range sites {
+		go checkStatus(site, 3, statuses)
+	}
+
+	results := make([]statusCheck, 0)
+	for i := 0; i < len(sites); i++ {
+		results = append(results, <-statuses)
+	}
+
+	emb := embed.NewEmbed().SetTitle("Website Statuses")
+	for _, result := range results {
+		title := ""
+		if result.Success {
+			title = "🆗 " + result.Site
+		} else {
+			title = "🔥 " + result.Site
+		}
+		emb = emb.AddField(title, fmt.Sprintf("Up: %v\nLatency: %dms\nError: %+v", result.Success, result.Time, result.Error))
+	}
+
+	s.ChannelMessageSendEmbed(m.ChannelID, emb.MessageEmbed)
+}
+
+// Check the up status of a website, returning an error if not up - checks if can connect and response is 200
+func checkStatus(site string, retryCount int, statuses chan statusCheck) {
+	var timeTaken int64 = 0
+
+	err := try.Do(func(attempt int) (bool, error) {
+		startTime := time.Now()
+		resp, err := http.Get(site)
+		timeTaken = time.Now().Sub(startTime).Milliseconds()
+
+		if resp != nil {
+			if resp.StatusCode != 200 {
```

Comment:
Looks good. What if the website returns a 307/308 for a www to non www redirect for example. Will it be registered as down?

Reply:
Hmm better to check against a list of possible error codes instead?

Reply:
Yeah I think so. Only ones i can think of off the top of my head are 200, 307 and 308

Reply:
>= 400 will prolly work

Reply:
Yeah I think so